### ISSUE TITLE
Syncbase Java: annoying to pass Type param to Put()

### ISSUE THREAD

**Issue by asadovsky** [vanadium/issues#888]
In Java, the signature of Table.Put() is `Put(key, value, type)`, unlike in all other languages, where it's `Put(key, value)`. Is it really necessary for the Java version of Put to take a type argument as a separate parameter? Why can't the type be derived from the value object, e.g. using reflection?
@yyoon

**Comment by kenashcraft** [comment]
@spetrovic77 @rogulenko any thoughts?

**Comment by spetrovic77** [comment]
We've had a discussion about this a while back, it comes down to what you
guys prefer to do for syncbase.  Below is the summary:

In Java, some types can't be deduced through reflection.  A simple example
is List<String> - the generic parameter "String" isn't visible to Java at
runtime, so we don't have enough type information to VOM-encode it
correctly.

Fortunately, all of the named VDL types have enough information to encode
without having to specify the extra type information.  I _think_ the same
holds true for user-defined basic classes, e.g.:

private static class Foo implements Serializable {
    private int i;
    private String s;

}

So, if we restrict ourselves to VDL types and these kinds of classes, we
could get rid of the type parameter.

I'm personally more in favor of having two versions of those methods: with
and without the type parameter, but I'd leave it up to you guys to decide.
Can the Syncbase query work over List<String> types (and bunch of other
types)?

-Srdjan

**Comment by asadovsky** [comment]
Do you think it would be too much to ask Java developers to call `Table.Put(key, ToVdlValue(javaObj, javaObjType))` if they wish to work with native types instead of VDL types?

This approach would avoids the need for having multiple versions of Put() (as well as any other such methods we might add, e.g. multi-put), and the behavior (in terms of what gets stored, and why queries might not work) is more transparent.

I'm not sure Syncbase queries will work with language-specific builtin types (@jakline will know better).

Also, note that in general we want to move towards encouraging developers to define VDL files for their schemas since that will enable stronger type checking, "update" queries like "update foo.bar.baz=3", secondary indexes on fields, and other schema-related features.

**Comment by spetrovic77** [comment]
The put(key, VdlValue) would be strictly worse in Java than even the existing solution, IMO.

Let's see what @jakline has to say, and then maybe just have the put(String key, Object value) method, where we make the best-effort to deduce the type for 'value' and fail if it can't be VOM-encoded.

**Comment by jakline** [comment]
> Can the Syncbase query work over List<String> types (and bunch of other
> types)?

Yes, if you mean a vdl value of a list of strings -- which should be produced by vdl when converting List<String>.  If you store a vdl Value, query will work with it.  That includes all vdl primitives (and Time) as well as structs and unions made up of all of the above.

Queries will not work at all if the value put is not a vdl.Value.

**Comment by spetrovic77** [comment]
In that case, we probably shouldn't limit Java functionality by prohibiting types like List<String> or Map<String, Blah> just to get rid of that extra type argument.

@asadovsky @kenashcraft Would you guys prefer if we had two versions of put() in java, one with and one without the type argument, or something else?

**Comment by asadovsky** [comment]
If it's uncommon to actually need the type argument, why not have a helper `ToTypedValue(value, type)` that users only have to call if their value happens to be one of those that requires passing a type?

With this approach, coupled with the fact that we plan to start encouraging use of VDL, it sounds like some high fraction ("99.9%") of users will get by with `Put(key, value)`, and that last 0.01% will get an error message, read our docs, and see that they should call `Put(key, ToTypedValue(value, type))`.

My inclination is to start with this approach because ToTypedValue is orthogonal to Syncbase, i.e. doesn't require an expansion of the Syncbase API surface. If we find it too cumbersome, it would be easy to add a `PutWithType(key, value, type)` in the future.

**Comment by spetrovic77** [comment]
Adam, we aren't doing ToTypeValue(Object), that's totally anti-Java.

Passing Type arguments isn't uncommon, and most commonly it will be:

put(key, value, Some.class);

**Comment by rogulenko** [comment]
Why don't you like to have both methods put(key, vdlValue) and put(key, value, type)? It's Java, we can overload.

**Comment by asadovsky** [comment]
+1, was just going to suggest the same. This way developers can provide the type argument as needed.
(Presumably put(key, value) can fail if it decides the type argument was needed, given the value?)

**Comment by rogulenko** [comment]
Yes, put(key, value) will fail for non-vdl values at compile time.

**Comment by spetrovic77** [comment]
Ah, that sounds better...

**Comment by spetrovic77** [comment]
So just to summarize, we'll have two methods:

put(String key, VdlValue value);
put(String key, Object object, Type type);

The first one will work for all VDL generated types, the second for everything else.  That sounds good to me.

**Comment by fredquintana** [comment]
The last proposal was to add a put(String key, VdlValue value) call to Table (and Row). This doesn't eliminate the need for the developer to specify the Type since the factory method for VdlValue takes (Object value, Type type). I think we need:

put(String key, Object value);
put(String key, Object object, Type type);

and then use reflection to determine the VdlType at runtime.
